src/environments/classic_control/acrobot.jl

```diff
@@ -0,0 +1,244 @@
+    action_space::DiscreteSpace{UnitRange{Int64}}
+    observation_space::MultiContinuousSpace{Vector{T}}
+    state::Vector{T}
+    action::Int
+    done::Bool
+    t::Int
+    rng::R
+    reward::T
+    # difference in second link angular acceleration equation
+    # as per python gym
+    book_or_nips::String
+    # array of available torques based on actions
+    avail_torque::Vector{T}
+end
+
+"""
+AcrobotEnv(;kwargs...)
+# Keyword arguments
+- `T = Float64`
+- `link_length_a = T(1.)`
+- `link_length_b = T(1.)`
+- `link_mass_a = T(1.)`
+- `link_mass_b = T(1.)`
+- `link_com_pos_a = T(0.5)`
+- `link_com_pos_b = T(0.5)`
+- `link_moi = T(1.)`
+- `max_vel_a = T(4 * π)`
+- `max_vel_b = T(9 * π)`
+- `g = T(9.8)`
+- `dt = T(0.2)`
+- `max_steps = 200`
+- `book_or_nips = 'book'`
+- `avail_torque = [T(-1.), T(0.), T(1.)]`
+"""
+function AcrobotEnv(;
+    T = Float64,
+    link_length_a = T(1.),
+    link_length_b = T(1.),
+    link_mass_a = T(1.),
+    link_mass_b = T(1.),
+    link_com_pos_a = T(0.5),
+    link_com_pos_b = T(0.5),
+    link_moi = T(1.),
+    max_torque_noise = T(0.),
+    max_vel_a = T(4 * π),
+    max_vel_b = T(9 * π),
+    g = T(9.8),
+    dt = T(0.2),
+    max_steps = 200,
+    seed = nothing,
+    book_or_nips = "book",
+    avail_torque = [T(-1.), T(0.), T(1.)]
+)
+
+    params = AcrobotEnvParams{T}(
+        link_length_a,
+        link_length_b,
+        link_mass_a,
+        link_mass_b,
+        link_com_pos_a,
+        link_com_pos_b,
+        link_moi,
+        max_torque_noise,
+        max_vel_a,
+        max_vel_b,
+        g,
+        dt,
+        max_steps
+    )
+
+    high = ([T(1.), T(1.), T(1.), T(1.), max_vel_a, max_vel_b])
+    env = AcrobotEnv(
+        params,
+        DiscreteSpace(3),
+        MultiContinuousSpace(-high, high),
+        zeros(T, 4),
+        2,
+        false,
+        0,
+        MersenneTwister(seed),
+        T(0.),
+        book_or_nips,
+        [T(-1.), T(0.), T(1.)]
+    )
+    reset!(env)
+    env
+end
+
+acrobot_observation(s) = [cos(s[1]), sin(s[1]), cos(s[2]), sin(s[2]), s[3], s[4]]
+angle_normalize(x) = ((x + pi) % (2 * pi)) - pi
+
+RLBase.observe(env::AcrobotEnv) =
+    (reward = env.reward, state = acrobot_observation(env.state),
+    terminal = env.done)
+
+function RLBase.reset!(env::AcrobotEnv{T}) where {T <: Number}
+    env.state[:] = T(0.1) * rand(env.rng, T, 4) .- T(0.05)
+    env.t = 0
+    env.action = 2
+    env.done = false
+    nothing
+end
+
+# governing equations as per python gym
+function (env::AcrobotEnv{T})(a) where {T <: Number}
+    env.action = a
+    env.t += 1
+    torque = env.avail_torque[a]
+
+    # noise to the force action
+    noise_range = env.params.max_torque_noise
+    if noise_range > 0
+        torque += T(2. * noise_range) * rand(env.rng, T, 1) .- T(noise_range)
+    end
+
+    # augmented state for derivative function
+    s_augmented = [env.state..., torque]
+
+    ode = ODEProblem(dsdt, s_augmented, (0., env.params.dt), env)
```

Comment:
I guess it doesn't really matter, if this environment isn't performance tuned.
But if you want to improve performance, you could replace `state` by `ode` in `AcrobotEnv` and set `ode.u0[1:end-1] .= ns` below and `ode.u0[end] = torque` above.